### PR TITLE
Change `dns_configuration_copy` to return remote value

### LINKED ISSUE #2713: Can't connect to Mongo on Node 20 + macOS
Node 20.17.0 + 20.16.0 doesn't work
Node 18 does work
Doesn't work both via CLI or IDE (WebStorm)
macOS
dotenv ^8.2.0
mongodb ^4.17.0
typescript 4.7.4
using `new MongoClient`

### PR DESCRIPTION
**Pull request by DmitryDodzin**
closes #2713

### CODE REVIEW

File: mirrord/layer/Cargo.toml [hunk 1]
```diff
@@ -49,6 +49,7 @@ exec.workspace = true
 syscalls = { version = "0.6", features = ["full"] }
 null-terminated = "0.3"
 base64.workspace = true
+resolv-conf = "0.7"
```
Comment (Razz4780) [inline mirrord/layer/Cargo.toml:52]: Should be moved to `[target.'cfg(target_os = "macos")'.dependencies]` section

Comment (DmitryDodzin) [reply mirrord/layer/Cargo.toml:52]: Correct

File: mirrord/layer/src/socket/hooks.rs [hunk 2]
```diff
@@ -442,68 +444,25 @@ pub(super) unsafe extern "C" fn sendmsg_detour(
...
-        pub reserved: [u32; 5],
-    }
-}
-
-#[cfg(target_os = "macos")]
-use macos::*;
-
 /// This implementation is actually enough for Netty case, since it seems to use the "standard"
 /// approach if resolver returned here is null TODO: return a real resolver based on remote
 /// resolv.conf
```
Comment (t4lz) [inline mirrord/layer/src/socket/hooks.rs:449]: That TODO is DONE, right?
Could you please also explain this detour in the function documentation?
What is this function supposed to do normally, what we do instead and why.

File: mirrord/layer/src/socket/hooks.rs [hunk 3]
```diff
@@ -442,68 +444,25 @@ pub(super) unsafe extern "C" fn sendmsg_detour(
...
-        n_scoped_resolver: 0,
-        scoped_resolver: std::ptr::null_mut(),
-        reserved: [0; 5],
-    }))
+    remote_dns_configuration_copy().unwrap_or_bypass_with(|_| FN_DNS_CONFIGURATION_COPY())
 }
 
 #[cfg(target_os = "macos")]
 #[hook_guard_fn]
 unsafe extern "C" fn dns_configuration_free_detour(config: *mut dns_config_t) {
```
Comment (t4lz) [inline mirrord/layer/src/socket/hooks.rs:458]: please document what this detour is for - what do we do different from the original and why.

File: mirrord/layer/src/socket/ops.rs [hunk 4]
```diff
@@ -1392,3 +1421,111 @@ pub(super) fn sendmsg(
...
+        let _ = Box::from_raw(nameserver);
+    }
+
+    let searchs = Vec::from_raw_parts(resolver.search, resolver.n_search as usize, 0);
+
+    for search in searchs {
+        let _ = CString::from_raw(search);
+    }
+
+    let _ = CString::from_raw(resolver.options);
```
Comment (t4lz) [inline mirrord/layer/src/socket/ops.rs:1481]: Why do these pointers need to be freed but not `domain` and `sortaddr`?

File: mirrord/layer/src/socket/ops.rs [hunk 5]
```diff
@@ -1392,3 +1421,111 @@ pub(super) fn sendmsg(
...
+}
+
+#[cfg(target_os = "macos")]
+#[mirrord_layer_macro::instrument(level = "trace", ret)]
+pub(super) fn remote_dns_configuration_copy() -> Detour<*mut dns_config_t> {
+    let remote = read_remote_resolv_conf()?;
+
+    // TODO: replace error
```
Comment (t4lz) [inline mirrord/layer/src/socket/ops.rs:1489]: If this TODO is meant to stay there, it needs more details.

File: mirrord/layer/src/socket/ops.rs (hunk 4 above)
Comment (DmitryDodzin) [reply mirrord/layer/src/socket/ops.rs:1490]: Well I only free any memory we only allocated, I changed so the or_bypass_with will return an empty one so the drop won't accidentally break or leak any memory

File: mirrord/layer/src/socket/ops.rs (hunk 5 above)
Comment (DmitryDodzin) [reply mirrord/layer/src/socket/ops.rs]: I changed it to be more detailed

File: mirrord/layer/src/socket/hooks.rs (hunk 3 above)
Comment (DmitryDodzin) [reply mirrord/layer/src/socket/hooks.rs:468]: Updated

File: mirrord/layer/src/socket/ops.rs [hunk 6]
```diff
@@ -1392,3 +1421,123 @@ pub(super) fn sendmsg(
...
+
+    let _ = CString::from_raw(resolver.options);
+}
+
+/// reconstruct a macos specific [`dns_config_t`] api from parsing the `/etc/resolv.conf` file from
+/// remote container and fill only the first resolver in the response
+#[cfg(target_os = "macos")]
+#[mirrord_layer_macro::instrument(level = "trace", ret)]
+pub(super) fn remote_dns_configuration_copy() -> Detour<*mut dns_config_t> {
+    let remote = read_remote_resolv_conf()?;
```
Comment (t4lz) [inline mirrord/layer/src/socket/ops.rs:1498]: shouldn't we at some point check the user's mirrord config? What if they disabled DNS?

Comment (DmitryDodzin) [reply mirrord/layer/src/socket/ops.rs:1498]: Ah

Comment (DmitryDodzin) [reply mirrord/layer/src/socket/ops.rs:1503]: Nvm the detours are not hooked when no dns

**Review by t4lz — APPROVED** [review]

**Pull request merged**